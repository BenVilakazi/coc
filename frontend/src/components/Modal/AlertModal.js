import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Button from '../Buttons/Button';
import Modal from './Modal';
import config from '../../config';

const propTypes = {
  bigText: PropTypes.string,
  smallText: PropTypes.string,
  buttonText: PropTypes.string,
  onClick: PropTypes.func,
};

const defaultProps = {
  bigText: 'SOMETHING WENT WRONG',
  smallText: 'An unrecoverable error occurred',
  buttonText: 'Click anywhere to restart',
  onClick: () => window.location.reload(),
};

const { smallDesktopWidth } = config.breakpoint.hostBreakpoints;

const AlertWrapper = styled.div`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  max-height: 180px;
  max-width: 360px;
  width: 90%;
  margin: auto;

  .main-box {
    display: flex;
    height: 100%;
    width: 100%;
    flex-direction: column;
    text-align: center;
    justify-content: center;
    align-items: center;
  }

  .top {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 100%;
    background: var(--secondary-background-color);
    color: var(--secondary-text-color);
    padding: 12px;
  }

  .bottom {
    display: flex;
    flex-direction: column;
    flex-grow: 3;
    justify-content: space-around;
    align-items: center;
    width: 100%;
    padding: 10px;
  }

  .button {
    background-color: var(--primary-background-color);
    color: var(--primary-text-color);
  }

  .button:hover,
  .button:active,
  .button:focus {
    background: var(--accent-background-color);
  }

  .big-text {
    font-size: 1.2rem;
    padding: 0;
  }

  .small-text {
    font-size: 1.2rem;
  }

  .button-text {
    font-size: 0.8rem;
    padding: 2px;
  }

  @media (min-width: ${smallDesktopWidth}) {
    max-height: 300px;
    max-width: 600px;
    width: auto;

    .bottom {
      padding: 12px;
    }

    .big-text {
      padding: 8px;
    }

    .small-text {
      font-size: 1.5rem;
    }

    .button-text {
      font-size: 1rem;
      padding: 8px;
    }
  }
`;

export default function AlertModal({
  bigText,
  smallText,
  buttonText,
  onClick: handleClick,
}) {
  return (
    <Modal onClickOutside={handleClick}>
      <AlertWrapper onClick={handleClick} className="primary-background-modal">
        <div className="main-box">
          <div className="top">
            <h1 className="big-text">{bigText.toUpperCase()}</h1>
          </div>
          <div className="bottom">
            <p className="small-text">{smallText}</p>
            <Button className="button">
              <p className="button-text">{buttonText}</p>
            </Button>
          </div>
        </div>
      </AlertWrapper>
    </Modal>
  );
}

AlertModal.propTypes = propTypes;
AlertModal.defaultProps = defaultProps;
