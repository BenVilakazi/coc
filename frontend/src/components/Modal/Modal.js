import React, { useRef, useEffect } from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const propTypes = {
  children: PropTypes.node.isRequired,
  onClickOutside: PropTypes.func.isRequired,
};

const Overlay = styled.div`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: rgba(0, 0, 0, 0.35);
  z-index: 9999;
`;

export default function Modal({ children, onClickOutside }) {
  const overlayRef = useRef();

  useEffect(() => {
    const handleClickOutside = (event) => {
      if (overlayRef.current && event.target === overlayRef.current) {
        onClickOutside();
      }
    };

    const handleKeyPress = (event) => {
      if (event.key === 'Escape') {
        onClickOutside();
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    document.addEventListener('touchstart', handleClickOutside);
    document.addEventListener('keydown', handleKeyPress);

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
      document.removeEventListener('touchstart', handleClickOutside);
      document.removeEventListener('keydown', handleKeyPress);
    };
  }, []);

  return createPortal(
    <Overlay ref={overlayRef} data-testid="overlay">
      {children}
    </Overlay>,
    document.getElementById('modal-root'),
  );
}

Modal.propTypes = propTypes;
