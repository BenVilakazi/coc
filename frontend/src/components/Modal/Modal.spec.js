import React from 'react';
import ReactDOM from 'react-dom';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Modal from './Modal';

afterEach(cleanup);

describe('Modal', () => {
  const handleClick = jest.fn();
  const handleChildClick = jest.fn();
  const consoleSpy = jest.spyOn(console, 'error');

  document.body.innerHTML = `
    <div id="modal-root"></div>
  `;

  // ----------------------------------------------------------------------------
  // rendering
  describe('rendering', () => {
    it('renders a child', () => {
      const { getByText } = render(
        <Modal onClickOutside={handleClick}>
          <div>hello</div>
        </Modal>,
      );
      expect(getByText('hello')).toBeTruthy();
    });

    it('renders multiple children', () => {
      const { getByText } = render(
        <Modal onClickOutside={handleClick}>
          <div>hello</div>
          <div>world</div>
        </Modal>,
      );
      expect(getByText('hello')).toBeTruthy();
      expect(getByText('world')).toBeTruthy();
    });

    it("does NOT render when children haven't been passed", () => {
      render(<Modal onClickOutside={handleClick} />);
      expect(consoleSpy).toHaveBeenCalled();
    });

    it("does NOT render when onOutsideClick prop hasn't been passed", () => {
      render(
        <Modal>
          <div>test</div>
        </Modal>,
      );
      expect(consoleSpy).toHaveBeenCalled();
    });
  });
  // ----------------------------------------------------------------------------

  // ----------------------------------------------------------------------------
  // events
  describe('events', () => {
    it('clicking the overlay triggers the callback', () => {
      render(
        <Modal onClickOutside={handleClick}>
          <div />
        </Modal>,
      );

      userEvent.click(screen.getByTestId('overlay'));
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('clicking the overlay does NOT trigger click events on children', () => {
      render(
        <Modal onClickOutside={handleClick}>
          <div onClick={handleChildClick} type="button" aria-hidden="true" />
        </Modal>,
      );

      userEvent.click(screen.getByTestId('overlay'));
      expect(handleChildClick).not.toHaveBeenCalled();
    });

    it('clicking children does NOT trigger the overlay callback', () => {
      render(
        <Modal onClickOutside={handleClick}>
          <div
            type="button"
            onClick={handleChildClick}
            data-testid="modal-child"
            aria-hidden="true"
          />
        </Modal>,
      );

      userEvent.click(screen.getByTestId('modal-child'));
      expect(handleChildClick).toHaveBeenCalled();
      expect(handleClick).not.toHaveBeenCalled();
    });

    it('fires onClickOutside when the escape key is press', () => {
      expect(handleClick).not.toHaveBeenCalled();

      render(
        <Modal onClickOutside={handleClick}>
          <div>Test</div>
        </Modal>,
      );

      fireEvent.keyDown(screen.getByText('Test'), {
        key: 'Escape',
        code: 'Escape',
      });

      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });
  // ----------------------------------------------------------------------------

  // ----------------------------------------------------------------------------
  // snapshot
  describe('snapshot', () => {
    // mock createPortal for snapshot
    beforeEach(() => {
      ReactDOM.createPortal = jest.fn((element) => element);
    });

    it('matches', () => {
      const tree = renderer
        .create(
          <Modal onClickOutside={handleClick}>
            <div>
              <h1>hello world</h1>
              <button type="button">click me</button>
            </div>
          </Modal>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
  // ----------------------------------------------------------------------------
});
