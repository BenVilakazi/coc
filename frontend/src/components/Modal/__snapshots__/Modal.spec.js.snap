// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal snapshot matches 1`] = `
.c0 {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: rgba(0,0,0,0.35);
  z-index: 9999;
}

<div
  className="c0"
  data-testid="overlay"
>
  <div>
    <h1>
      hello world
    </h1>
    <button
      type="button"
    >
      click me
    </button>
  </div>
</div>
`;
